Review (#2)
marlonschlosshauer/thesis: Thesis.org

#+CAPTION: Example of parallel composition of a sign-in component. The first composition is the result of adding a button and two inputs. We further compose the first composition by adding a headline. That result is then again further composed by adding a logo.
#+ATTR_LATEX: :width 350px
[[./images/parallel-composition-highlighted.png]]

However, modern frontend applications have gone beyond just serving information with HTML. Websites often act as a digitalization of a workflow, where the order of components, as in which component is shown after another, is crucial. This is also a kind of composition. Instead of composing things in parallel, we're composing sequentially here. That is to say, we say which steps we have and in what order they appear. Together they form a sequence.

> The resulting sequentiality is different from how we know it from imperative programming, as we don't mutate state with each step. Rather the result of each step is fed into the next. This allows us to address the essential sequentiality of the domain while not creating accidental sequentiality, as we still have access to the tools provided by functional programming.

Reviewer: The difference between sequentiality in imperative programming (IP) vs. sequential composition in our case needs to be more precise. My take: Sequentiality in IP is a means for the implementation of ideas that aren't necessarily sequential in essence. Sequentiality in IP is therefore accidental, it's an accident of history that some programmers still program imperatively. In contrast, the kinds of sequentiality we want to deal with are naturally sequential. We want to express this essential/natural sequentiality in code. Example: In IP you have to program the summation of a list of numbers sequentially (with a loop) even though "summation" as an abstract idea doesn't involve sequentiality at all. On the other hand, the task "We need a workflow of two tasks, one following the other" is naturally sequential and we therefore want to express this sequentiality in code, e.h. `(then task-1 task-2)`.

Author: ```suggestion
It is important to point out that the here mentioned sequentiality is different from how we might know it from imperative programming (IP). In IP, sequentiality is used, even if the domain doesn't demand it, because of its writing style. We could call this accidental sequentiality, as it is only a by-product of language design and history, rather than a conscious engineering decision.
Because our sequential composition exhibits natural sequentiality (e.g first step ~A~, then step ~B~) we also want to represent that in our code.
```

I tried making the differentiation between accidental and natural clear this time. Maybe I should also mention about shortcomings of accidental seq? From my current understanding that would be more code and more error-prone code. Is that correct?

Reviewer: Very good!

> Maybe I should also mention about shortcomings of accidental seq?

If you have the time. I don’t think that’s absolutely necessary.

#+CAPTION: Example of sequential composition of a sign-up component. The first composition combines two steps, before being again further composed by adding a third step.
#+ATTR_LATEX: :width 350px
[[./images/sequential-composition-highlighted.png]]

Because of this dependency, composing sequentially isn't just an optimization. It is a fundamental building block of how an application is supposed to operate. Just like how we need the ability to show both an input field and a button at the same time (to confirm the input), we also need the ability to model what happens after (and possibly before) that button has been pressed.

But despite that need, the actual support for sequential composability by modern frameworks is either non existent or very poor.
While some of the desired result can be achieved by making clever use of parallel composability or using traditional links, most aren't truly sequentially composable and both suffer from multiple issues, some of which are:
- They don't yield things which we can compose further (into more sophisticated workflows)

> - Reusing these components leads to a lot more abstraction

Reviewer: I'd say it "leads to more boilerplate/scaffolding". "More abstraction" or rather "better abstraction" is what you try to achieve with this thesis.

Author: ```suggestion
- Reusing these components leads to a lot more boilerplate
```
Yep 👍

- It produces code that becomes difficult to understand at a glance (or at all)

> - Most importantly: the logic for advancing the steps and the book-keeping lives now inside the component, with the parallel composing logic

Reviewer: I don't get this point. Isn't the problem that all the book-keeping is scattered throughout the codebase?

Author: ```suggestion
- Logic for advancing steps (and bookkeeping) live besides the parallel composing logic
- Bookkeeping is scattered throughout the codebase
```

I'd say having both at the same place could cause bugs (accessing bookkeeping/state incorrectly) but that does seem unlikely. "Booking scattered around" is a big problem and I would add that to the list.

There are more issues. The inability to easily test just the order of the components and the huge time cost required to refactor constructs of this nature are just the starting point.

> Our applications already are a collection of parallel compositions. If we add the ability to compose sequentially, we gain another tool to control our programs. Not only would our code be more expressive and more concise, bugs could also be reduced because the resulting systems represent more truthfully what they set out to do, thanks to the provided API of the framework. Most importantly, we could build workflows out of smaller flows or other workflows entirely and share these across our applications, just how do with our UI elements.

Reviewer: "because the resulting systems represent more truthfully what they set out to do" yes, very good!

Another important feature of React is that the developer must explicitly declare the state of a component. This is used to improve performance, as only components which are effected by a state change are rendered upon changes. Old components remain untouched. This also helps with identifying where state is and how it might be changed in the future.

React is also a popular basis for other frameworks like Preact. It is especially popular for ClojureScript based frameworks. Noteable examples are reacl-c, om, reagent, rum and others.

> * What is reacl-c

Reviewer: The key advantage of reacl-c is that its API is denotational, i.e. its building blocks have a concise semantics and the semantics of combined items follows the semantics of the parts. (Just as an aside. Probably not relevant for your thesis.)

Author: I'd love to expand this this chapter by also talking about how reacl-c delivers exactly what devs tell it to. It's on my TODO list.

Reacl-c [2] is a UI framework for ClojureScript that is designed to make the composition of UI components easier. Like other ClojureScript frameworks, reacl-c wraps React to best take advantage of Clojure's immutable data structures. Like React, reacl-c allows the definition of components, called ~items~. These ~items~ can have child ~items~, thus forming a tree, just like components do in React.

> To make composing of ~items~ easier reacl-c takes a different approach to state than React. There are two types of state. Local state, which is defined inside of the component and outter state, which is implicitly passed down from the parent to the child. The later has the unique feature of putting the emphasis on the parent, in their relationship. We will explore why this is important soon.

Reviewer: "outter" -> outer

Another key difference to React, is that ~items~ can emit actions if an event occurs inside of them. These actions propogate upward the item tree. Every item can register an action handler, which captures the action and reacts to it. This enables the developer to define a (child) ~item~ completely independently from their potential future parents, leading to more reusable and more composable components.
Instead of having the parent pass down a function to, e.g manipulate state, the component can emit an action upwards and trigger the same functionability in the parent. Another benefit is that the parent component can adapt the output of the emitted action further.

#+CAPTION: Showcasing the inversion of control by comparing emitting of actions to passing down callbacks. Each circle represents a component. To communicate the callbacks need to be passed down. With actions, the children can speak up on their own.
#+ATTR_LATEX: :width 350px
[[./images/emit-vs-callback.png]]

> As an example: A button toggles an option flag from ~true~ to ~false~ and back. The state for this option is put into the parent, our button is a toggle component which will be reused across the application. In React, the button needs to receive a function which to callback, after the button has been pressed. In reacl-c the button does not need to receive anything from the outside world, as the button emits an action when pressed, which the parent can capture and act upon.

Reviewer: The action system in reacl-c is inspired by so called `effect systems` in FP. Maybe you want to drop that fact and add another item to your references.

Author: ```suggestion
As an example: A button toggles an option flag from ~true~ to ~false~ and back. The state for this option is put into the parent, our button is a toggle component which will be reused across the application. In React, the button needs to receive a function which to callback, after the button has been pressed. In reacl-c the button does not need to receive anything from the outside world, as the button emits an action when pressed, which the parent can capture and act upon.

The action system in reacl-c is inspired by the functional programming concept of effect systems. An effect system allows code to express effects, by tracking them directly through the type system[x]. Like an effect system, the action system in reacl-c allows us to express these effects by returning either the ~Action~ or ~Return~ type.  
```

Great idea. I added a small explanation. Is my understanding of an effect system correct? I'm OK with the length, but if you have something to add I'll gladly expand it. I think I could mention that ``commit`` is also an effect type.

I took the explanation from  [this paper](https://macau.uni-kiel.de/servlets/MCRFileNodeServlet/macau_derivate_00003380/kcss_2021_07.pdf#page=85) (page 85, "Background").

Reviewer: Yes, that’s accurate.

To make components even more composable the parent can not just control the result, it can also control which state is given to its children in the first place. The parent can therefor handle the child items like they're pure functions, that together with the parent compose to a new item.

In React often another package like Redux is used to fix the problem of having to pass-down functions. Instead of saving all state in each component, state can be managed at a central point which components can send messages to, causing the central state to change. This can cause issues with the composability of components. The problem with this approach is that components cannot be placed multiple times into the app, without making sure they don't all work on the same central state, first. Reacl-c solves this issue by allowing any component to send and receive messages. Not only does this allow for local reasoning, but it also enables us to wrap a component with an action handler and placing it infinitely in the app without worry, as the component won't effect anything outside of that handler.
** How an Item is made

> Reacl-c is made up of so-called ~items~ and functions which operate on these ~items~. Much like in React, these ~items~ are instructions that represent HTML elements that can be nested and can hold state.

Reviewer: Mhh. "instructions" is an incorrect description. Furthermore, some items don't map to HTML elements at all (such as network requests).

Author: ```suggestion
Reacl-c is made up of ~Items~ and functions which operate on these ~Items~. Much like in React, we can use these ~Items~ to build our components.
The ~dom~ namespace offers all the necessary HTML elements in the form of ~Items~ (e.g ~dom/div~, ~dom/button~). We are however not limited to HTML or even visible elements. It is also possible to place empty ~Items~ to cause effects (e.g HTTP request).
As an example, to create a headline all that is needed is ~(dom/h2 "Text")~. To make more complicated ~Items~, combinators like ~(dom/div)~ or ~(core/fragments)~ can be used. These can contain multiple ~Items~. See listing 3 for an example.
```

Yea "instructions" makes it seem like imperative prog. I cut it out and now mention empty items, too.

To make an ~item~ reacl-c provides a namespace called ~dom~ that contains all the necessary DOM HTML components already defined as ~items~. As an example, all that is needed to create a headline is ~(dom/h2 "Headline")~. To make more complicated ~items~, combinators like ~(dom/div)~ or ~(core/fragments)~ can be used. These can contain multiple items. See listing 3 for an example.
#+CAPTION: Showcasing composition of multiple ~items~ into one by using a ~div~.
#+begin_src clojure
  (dom/div
   (dom/h2 "Hello")
   (dom/p {:class "subtitle"} "This is some text")
   (dom/button "Subscribe"))
#+end_src
There are other functions which, much like ~div~, don't add something visually but change the behavior of the ~item~. Functions like ~core/focus~, ~core/dynamic~ and ~core/handle-actions~, to just name a few. We will take a closer look at each later.

> ~Items~ can also receive a ClojureScript map as their first argument. With this map things like CSS classes and inline-style can be applied. If the ~item~ is interactive, like Buttons and inputs are, the keywords ~onChange~ and ~onClick~ can be used to register a callback.

Reviewer: `dom`-items can receive […] a map . Strictly speaking no item can. Something like `dom/div` is an item-constructor. A function that takes said map and a list of children and returns an item.

Author: ```suggestion
Behind the calls to the ~dom~ namespace are ~Item~ constructors, which can also receive a ClojureScript ~map~ as their second argument. With this ~map~ things like CSS classes and inline-style can be applied. If the ~Item~ is interactive, like Buttons and Inputs are, the keywords ~onChange~ and ~onClick~ can be used to register a callback.
```

I don't mention the list of children, but I think it's enough info. What do you think?

Reviewer: Sounds good!

** Emitting and handling actions
Actions are an important tool for children to communicate with their parents. Like the implicit (or explicit) passing of state from the parent, this is a crucial feature to make items composable.
To emit an action the child needs to call ~core/return~ with the ~:action~ keyword and the desired payload. The action will travel upwards until it is caught by a ~core/handle-action~. The action handler function takes two arguments. First an ~item~ to wrap and receive actions from. Second, a function that defines what do to when it receives an action.
#+CAPTION: The action emitted by the buzzer ~item~ is caught in the parent ~item~ using ~core/handle-action~.
#+begin_src clojure
  (def buzzer
    (dom/dom
     (dom/h3 "Press buzzer to notify parent")
     (dom/button {:onClick (fn [] (c/return :action true))} "Buzzer")))

  (def parent
    (core/local-state
     {:pressed? false}
     (core/handle-action
      (core/dynamic
       (fn [[outter inner]]
	 (dom/div
	  (dom/p (str "The button has" (if (:pressed? inner) "been pressed" "not been pressed")))
	  buzzer)))
      (fn [ac msg]
	(core/return :state {:pressed? msg})))))
#+end_src

> It is important that the function which is given to ~core/handle-action~ calls ~core/return~ at the end to either notify a parent or itself of the state change, as an action is a side-effect.

Reviewer: No, the default, i.e. returning `nil` is just going to suppress all actions which leads to bugs. Actions aren't "side-effects" in the logic of reacl-c. They do have an effect on the system but they're not "side-effects" in the usual sense. Maybe just scrap that whole paragraph.

Author: ```suggestion
```
👍

** How state is managed
Like with React, handling state is very important in reacl-c. The framework gives the developer many ways to tackle this problem. Much like with React, developers can easily make out if a component is using or changing state, which makes working with state easier.

State in reacl-c can be shared in different, more complex, ways.
Firstly, while a component might have state, it is not accessible to the developer until they use the ~core/dynamic~ function or ~core/with-local-state-as~ macro. This has the benefit of instantly marking a component as one that needs and works with state.
Secondly, like mentioned earlier, state is split into two categories:
- Inner state, which is defined inside of the component by using either ~core/with-state-as~, ~core/local-state~ or ~core/isolate-state~.
- Outter state, or state that is passed down from the parent component.

> The inheritance of state from the parent happens implicitly, but can be controlled by the parent through lenses (with the ~core/focus~ function). These tools are a reason why reacl-c ~items~ are so composable.

Reviewer: "a reason why reacl-c items are so composable" needs more justification, but that's probably our (the authors of reacl-c) responsibility instead of yours. Let me try:

State being "passed implicitly" sounds strange at first. How does that make items "composable" and not just more confusing, as is often the case with implicit stuff happening? We tackle this question when we look at how "return values" are passed implicitly in functional expressions as well:

`(inc (+ 3 4))`

Here the result of the subexpression `(+ 3 4)` is "passed" to the input of `(inc ...)` "implicitly", i.e. without saying something awfully verbose like "Let there be two computations, the first being the addition of two numbers, the second being the incrementation of a number. And then let us take the output of the first computation to be the input of the second." No, we just say `(inc (+ 3 4))`. As a reader of such an expression, we "just know" (via education) that in purely functional programming every expression has a value and we "just know" that expressions can be substituted by their values without altering semantics.

This is akin to the implicit passing of state in reacl-c. Instead of saying "Let there be two components in our hierarchy and the allocation of state proceeds like this: …", we say: `(div (focus lens/first input-text) (focus lens/second input-text))`. We "just know" that each item manages state. `focus lens/second` transforms our (the reader's) stance towards state like `inc` transforms our stance towards return values.

Author: ```suggestion
That passing down of outer state happens implicitly. That means we don't need to tell the parent to share its state with its children. 
To understand this we need to look at an example. If we write ~(inc (+ 3 4))~ we specify that the result of ~(+ 3 4)~ should be passed to ~(inc ...)~, but we don't do so in a verbose way. Rather it is implicitly done because of the structure of the code.
We know that having functions nested inside of each other causes this kind of passing of return values. ClojureScript developers are already familiar with this, as it is arguably the most important feature of the language.

Let's apply similar thinking to state and ~Items~ in reacl-c. Take a look at the following: ~(div (core/local-state {:text "Text"} input-text))~. We don't have to explicitly say that ~input-text~ should inherit the state (e.g ~{:text "Text"}~) from the parent. The structure indicates this relationship.
If we take our example further it becomes clear why our two examples are similar. If ~input-text~ has access to ~:text~, it can then change it. If it does, it is like ~input-text~ returns the result of some computation to the parent (because the state belongs to the parent), where it is then used for further computation.
After all, we placed ~input-text~ in the parent for a reason. In this case to return something to us. Just how we placed ~(+ 3 4)~ inside of the ~(inc ...)~ to return something, as well.
In both cases we didn't specify that relationship in a verbose way (e.g like with a ~let~), but rather it was implicitly known to the developer by the relation.
It's worth noting that the parent retains full control of what state is shared, as it can use lenses (with the ~core/focus~ function) to block some of its state from being shared with its children.

It does not only lead to less code that needs to be written (because of the absence of all the wiring) but also has an effect on composability, because now children can be placed within anything. State is implicitly fed into ~Items~ and changes automatically find their way up to the parent (and their parent, and the parent of that parent etc.).

If state inside of the parent isn't in a format that is useful to the child, it can be further composed by wrapping it in another component that adapts the parent state by transforming it into the correct format. This is again possible because state is also implicitly passed into the adapter and from the adapter into the actual child. We will explore this mechanism in-depth later.
```

Thanks for the in-depth explanation! I now go deeper into why this makes sense ("not just more confusing"). I took your example of ``inc`` and ``(+ 3 4)`` and put more emphasis on the "relation" causing the implicit behavior. 

Is the comparison between the two (both returning a "result"), that I draw, correct? 

I had trouble coming up why **this** makes it more composable (last paragraph). If you have more on that I'd love to add it.

Explaining it with "we just know" is good. Not sure how I add that without sounding redundant. Do you think its clear enough here?

Reviewer: > We know that having functions nested

having function _calls_ (or _applications) nested

> the most important feature of the language.

one of the most important features of the language and the lambda calculus in general.

> Do you think its clear enough here?

I think so, but you’d probably have to show these paragraphs to an outsider. An outsider can judge whether the explanation is good enough. The both of us know too much already ;-)

To be a monad the type needs to provide two functions and satisfy three rules[4]. The required functions are ~>>=~ (also called ~bind~) and the ~return~ function.
A ~bind~ takes an instance of a monad ~M~ and a function that gets a value ~a~ and returns an instance of type ~M~ with ~a~ inside of it. The result will be a monad ~M~ again. This is what enables us to chain these operations together.
The second function, the ~return~, takes a value ~a~ and returns a monad ~M~ with value ~a~. As an example ~Just 1~ works like a return, in that we give it a ~1~ and it gives us a ~Maybe~ (with the value of ~1~ inside of it). The ~return~ function is sometimes also referred to as ~pure~. To better understand the signature, see listing 7 where Haskell notation has been used to spell out the types.
#+CAPTION: Haskell notation of the functions ~bind~ and ~return~
#+begin_src haskell
  M a >>= (a -> M b) = M b
  return a = M a
#+end_src

An implementation of these functions needs to fulfill the following three rules to be considered a monad[5]:
- Left identity: ~return a >>= h = h a~
- Right identity: ~m >>= return = m~
- Associativity: ~(m >>= g) >>= h = m >>= (\x -> g x >>= h)~

> Left and right identity are tests to make sure the types work out correctly. ~return~ can both be called with a value to create a monad, when provided on the left side of the bind, or be given as a ~continuation~ function, if provided on the right side. The rule of associativity tests that the order of operation is not important. Both ~(A >>= B) >>= C~ and ~A >>= (B >>= C)~ should yield the same result.

Reviewer: The "order of operation" is important. It's, well, the association of operations that's unimportant.

Author: ```suggestion
Left and right identity are tests to make sure the types work out correctly. ~return~ can both be called with a value to create a monad, when provided on the left side of the bind, or be given as a ~continuation~ function, if provided on the right side. The rule of associativity tests that the order of operation remains, even if the association of the operations changes. Both ~(A >>= B) >>= C~ and ~A >>= (B >>= C)~ should yield the same result.
```

👍

Most importantly, we again mix our parallel and sequential composition. We want the logic of both to be seperated, because they do different things. Yet every parallel component now needs to do bookkeeping as well.
Lastly, besides the parameter in the signature, there is no way to distinguish from the outside if this is a sequential or parallel component. Placing a single step of a sequential composition into a parallel composition might not make sense and we'd like to make this not possible on an API level.
** Conclusion
Current methods either don't offer composability or they require a lot of boilerplate to be composable. It is possible to write functions which abstract that boilerplate code away, but that takes time and effort.
Support from modern frameworks could not save developers time but also deliver smooth integration with other features of the framework.
* Desired Behaviour
The goal is to create both an easy to use yet powerful tool to compose sequentially. Lets first look at some required features that our implementation should deliver before we talk about how a developer might interact with it.

> Our sequential composition needs to handle the synchronicity of the domain. Only one value is to be shown to the user at a time and that is the current value in our list of ~key:value~ pairs. Execution of pairs further in the list needs to be halted until they're displayed. The same should hold true for HTTP requests which a developer could place in our composition.

Reviewer: The following paragraphs talk about syntactic (and therefore secondary) concerns. Everything the `runner` macro does can be achieved with pure `bind` and `return`. Maybe introduce a new section header?

Author: I added a comment below with a suggestion that adds two headings and talks more about semantics 👍

#+begin_src clojure
  (runner [personal (personal-info)
	   verification (verification-code personal)]
	  (show-account-info [personal verification]))
#+end_src
A possible alternative to this would be to omit the body function entirely and instead use the element previously placed in the body (~show-account-info~ in our example), as the last element in our ~key:value~ pairs list. While not commonly used like this, ~let~ also allows for this style. This has the benefit of creating a concise and consistent look and feel.
#+CAPTION: Login example in ~let~ style without a body function
#+begin_src clojure
  (runner [personal (personal-info)
	   verification (verification-code personal)
	   info (show-account-info [personal verification])])
#+end_src
The ~runner~ function executes our composition and should be able to be used just like a regular ~item~ when wanted. It should be further composibly in parallel with other ~items~ and actions should be able to be caught from it.
Inside of the ~runner~ we will have steps. Those steps should be composable thanks to another function. For now we can steal terminology from the earlier introduced monads and call this function ~bind~ or ~then~.

> These two functions should be the primitives of our implementation.

Author: ```suggestion
** What it should do
Our sequential composition needs to handle the synchronicity of the domain. That means only one value is to be shown to the user at a time. It's often the case that a step depends on information provided in the previous step. As such it is important for our composition to allow future steps to access the values returned by earlier steps.
Because of similar requirements with synchronicity and order, it should also be possible to compose other asynchronous operations such as HTTP requests.
Lastly, individual steps should be fully composable. They should be shareable and have the ability to nest in complex ways.
** How a developer should be able to use it
Unlike parallel composition, there are little examples to go off of. Because of this it makes sense to draw inspiration from other, already well established, functions.
ClojureScript provides something that allows us to define things in a neat way with the ~let~ function[6]. As seen in listing 13, a ~let~ is composed of two parameters. The first parameter is a list of ~key:value~ pairs, where a ~key~ is nothing but a name for a symbol that will be used within the ~let~ and ~value~ is the actual value of that ~key~. The second parameter, also called body, is a function which has access to the previously defined keys. It will be run once all the ~values~ have been computed.
#+CAPTION: Using ~let~ to bind values to the names ~one~, ~two~, ~three~. The body is the last line, which is an operation that returns ~6~.
#+begin_src clojure
  (let [one 1
    two (+ 1 one)
    three (inc (* two one))]
    (+ one two three))
#+end_src

Adopting this style for sequential composition has many benefits. The most obvious is that ClojureScript developers would already be familiar with it. Even reacl-c offers a ~let~ style macro in the from of ~ref-let~[7]. It also satisfies a lot of our requirements, like being able to access prior results and only executing one pair at a time. Listings 14 shows how our earlier sign-up example could look like if we used a ~let~ like style. Note how little code is needed. This example still closely follows ~let~ in that it needs an uneven amount of arguments, in which the last is a function that will be executed (with access to all the previously declared ~keys~) at the end.
#+CAPTION: Earlier sign-up example written in a ~let~ style
#+begin_src clojure
  (runner [personal (personal-info)
       verification (verification-code personal)]
      (show-account-info [personal verification]))
#+end_src
A possible alternative to this would be to omit the body function entirely and instead use the element previously placed in the body (~show-account-info~ in our example), as the last element in our ~key:value~ pairs list. While not commonly used like this, ~let~ also allows for this style. This has the benefit of creating a concise and consistent look and feel.
#+CAPTION: Login example in ~let~ style without a body function
#+begin_src clojure
  (runner [personal (personal-info)
       verification (verification-code personal)
       info (show-account-info [personal verification])])
#+end_src
The ~runner~ function executes our composition and should be able to be used just like a regular ~Item~ when wanted. It should be further composibly in parallel with other ~Items~ and actions should be able to be caught from it.
The actual sequential composition will be done using the ~bind~ function (to compose things further) and ~return~ (to create a composible type). Developers can use these functions to build up their sequential compositions outside of the ~runner~. The ~runner~ itself will act as an abstraction to those two primitive functions.
```

** Used types
While ClojureScript is a dynamically typed language, it is helpful to create types using Clojures ~records~ to make handling and transforming data easier.
The most fundamental type is an UI element, which reacl-c already supplies in the form of ~Item~. To signal that the next step should be executed the ~Item~ needs to emit something which can be recognized internally. For that purpose the ~Commit~ record exists. If an ~Item~ emits a ~Commit~, the internals will execute the next step.
A developer could just pass an ~Item~ for composition, however, it makes sense to have the developer acknowledge that they're working with more than just a simple ~Item~. After all, the ~Item~ should at some point emit a ~Commit~ to change the currently shown step. So, to be able to use the ~item~ for sequential composition, the developer needs to wrap it in a ~Prog~. This signals that the developer understood that the ~Item~ will eventually emit a ~Commit~.
In short:
- ~Item~: UI element
- ~Commit~: What an ~Item~ emits to signal that the sequence can continue
- ~Prog~: An ~Item~ that will ~Commit~

> - ~Bind~: Product of composing a ~Prog~ further. Can be composed with itself further.

Reviewer: It's important to note that `Bind` is a subtype of `Prog`: every bind is a prog. This is what makes Progs endlessly composable.

Adding `Bind` as a type is not very helpful. It's more like a value constructor that makes a Prog: `bind : Prog -> Cont -> Prog` or more detailed `bind : Prog a -> (a -> Prog b) -> Prog b`

Author: Calling it a value constructor makes it a lot more clear. I added a suggestion in the comment below. Do you think I need to add more or is it clear enough?

> Internally, there is also a ~Bind~ type, which is the result of a ~then~ call. This holds both a ~Prog~ and a ~continuation~. The continuation will be called once a ~Commit~ has been captured from the ~Item~ inside of the ~Prog~.

Author: ```suggestion
- ~Prog~: An ~Item~ that will emit a ~Commit~
Internally ~Prog~ has a subtype called ~Bind~, which is the result of a ~then~ call (see next chapter). The ~Bind~ holds both a ~Prog~ and a ~continuation~. The ~continuation~ will be called once a ~Commit~ has been captured from the ~Item~ inside of the ~Prog~. Because it is a subtype, every ~Bind~ is also a ~Prog~. This will enable us to endless compose ~Progs~ with ~then~.
```

** Functions exposed by the API
To deliver on the promises of frictionless composability without loss of performance, monads are used. Because of that, the API needs to provide the ~return~ and ~then~ (also called ~bind~) functions to be considered a monad. Further, to display a ~Prog~ or ~Bind~ easily, a ~show~ function has been added. The most important function is ~runner~, which executes a ~Prog~ or ~Bind~ inside of it, allowing it to walk through the provided steps.
The primivate functions are ~return~, ~then~ and ~runner~.
*** return

> The ~return~ function takes an ~Item~ and turns it into a ~Prog~. This allows to go from a parallel composition (with an ~Item~) to a sequential composition (of a ~Prog~). Once an ~Item~ is a ~Prog~ the result can't be further parallely composed.

Reviewer: This doesn't sound like the correct semantics for `return`. Can't you toss arbitrary values to `return`, like `(return 123)`?

Author: Right now ``return`` only takes ``Items``. You can return arbitrary things through commits. If I remember correctly the problem is that ``(return 123)`` will never trigger the ``handle-action`` inside of the runner. I'm sure I could change it to allow for it though.

Reviewer: For the entire machine to be a "monad", we need a `return`-like function, that works for arbitrary values. Probably your `return` is not the `return` of the monad, then.

The monad `return` has to relate to `bind` just like a `0` relates to `+`, such that `bind (return x) f == f x` for all x of any type.

Reviewer: Maybe a less abstract example. We’d like to write a program such as this:

(defn get-an-answer []
  (bind (some-network-call)
            (fn [result]
               (if result (return result) (return "some default answer")))))

Author: > such that `bind (return x) f == f x` for all x of any type.

That makes a lot of sense. The types also say e.g ``M a`` and not ``M a (where a is subtype of b) blabla``. I'll add it after out char tomorrow   👍

*** then

> The ~then~ function is what allows us to compose multiple ~Progs~ together. For that it takes both a ~Prog~ and a ~continuation~ function (which should return another ~Prog~). The ~continuation~ will be called later, in the ~runner~ function. ~then~ creates a value of type ~Bind~ by passing the ~Prog~ and ~continuation~ parameters along. To allow for composition, the developer can also pass a ~Bind~ instead of a ~Prog~. The ~Bind~ in that case would be a previously created composition.

Reviewer: Here you haphazardly explain the concept of a subtype Bind < Prog. Mention the subtype relation above, then this paragraph doesn't have to mention it again.

Author: ```suggestion
The ~then~ function is what allows us to compose multiple ~Progs~ together. For that it takes both a ~Prog~ and a ~continuation~ function (which should return another ~Prog~). ~then~ actually creates a ~Bind~ (subtype of ~Prog~). The ~continuation~ will be called later, in the ~runner~ function.
```

Trimmed it. Kept the mention of ``Bind`` as the result, so it's documented, but I can remove it if you think it's unnecessary.

It takes a single ~Bind~ (or ~Prog~) as an argument, which could contain further ~Binds~ inside of it. Once a ~commit~ is emitted from the ~Bind~ that it displays, it calls the ~continuation~ of the ~Bind~ and displays the result of that ~continuation~. If the result is another ~Bind~, emitting another ~commit~ will trigger a call to the ~continuation~ of the new ~Bind~, which should produce yet another ~Bind~ etc.
*** show
~show~ extracts the ~Item~ from the passed parameter, allowing it to be displayed. If it's a ~Prog~ it just takes the ~Item~ inside of the ~Prog~ and displays it. If it is a ~Bind~, it first takes the ~Prog~ inside, then shows the ~Item~. If an ~Item~ is passed, the same ~Item~ will be returned. Show serves as one of two ways to turn a sequential composition back into a parallel one. This however does not capture any emitted ~commits~. If the execution of sequential composition is desired, ~runner~ should be used instead.
** Making the API more intuitive with macros

> Earlier we discussed a possible implementation for our sequential composition, using ClojureScripts ~let~ as an inspiration.

Reviewer: Aha! I think mentioning the `let`-style syntax above ("Desired Behaviour") adds no value. In fact, I was very confused on the first read. This section right here conveys your intents a whole lot better.

Author: I see "Desired Behaviour" more as an introduction. To be able to compare "current vs future". 
My intention was to first give a syntactic example how it could be done and to later dive deeper into the semantic and implementation. Do you think that is still possible or should "Desired Behaviour" just be axed?

We choose that function, among other things, because it's ~key:value~ structure is well understood and is much easier to read than nested anonymous function calls. However, to achieve our goal we need to make use of a ClojureScript feature called macros.
*** Why are macros used here?
From now on, the previously defined functions like ~then~, ~return~, ~runner~ will be called primitives. While these enable us to compose sequentially, they don't work like ~let~ does. Listing 17 shows the current usage of ~then~.
#+CAPTION: Sequential composition using primitive ~then~ with anonymous function calls.
#+begin_src clojure
  (then first
	(fn [a]
	  (then second
		(fn [b]
		  (then third
			(fn [c]
			  (done [a b c])))))))
#+end_src
The developer needs to create the anonymous functions by hand, everytime. Thankfully, the functionality of binding ~keys~ to ~values~ stays the same, as results of the components, like ~first~, will be bound to the parameter in the ~continuation~ (e.g the symbol ~a~).
Because this is functionaly identical, we can write a macro to translate the code example above into the desired ~let~ like structure.
*** What are macros?

> Macros are a powerful feature which lets us rewrite our ClojureScript code before it's being evaluated. We do this by selectively telling the language what to evaluate and what to keep for later evaluation. That allows us to use all of ClojureScripts functions to manipulate the input code. This is made possible partly because ClojureScript is a Lisp, so the code already looks like a Clojure data structure. The language uses this to its advantage to operate on itself. The return value of a macro will be a list of code, that will then be evaluated. Because of this we can use the entire language to transform our code, like we transform data, into something more usable.

Reviewer: "We do this by selectively telling the language what to evaluate and what to keep for later evaluation." Confusing. Can you elaborate? I'd say that the preceding sentence is sufficient. Macros treat code as data.

Author: ```suggestion
Macros are a powerful feature which lets us rewrite our ClojureScript code before it's being evaluated. That allows us to use all of ClojureScripts functions to manipulate the input code. This is made possible partly because ClojureScript is a Lisp, so the code already looks like a Clojure data structure. The language uses this to its advantage to operate on itself. The return value of a macro will be a list of code, that will then be evaluated. Because of this we can use the entire language to transform our code, like we transform data, into something more usable.
```
Shorter = Better 👍

Macros can be found all over Clojure and ClojureScript. Functions like ~when~[8] and ~when-not~[9] are actually macros that rewrite themselves to a simple ~if~. Even the much discussed ~let~ function is a macro[10].
*** Syntax
The ability to delay evaluation of just some parts of our code is powerful. ClojureScript gives a couple of new symbols to decide how it should evaluate code. We can use these symbols inside of a ~defmacro~ function, which works similar to ~defn~ and ~def~ with which we define functions and values with.
The following table lists the symbols:
| Symbol | Name           | Function                          |
| '      | Quoting        | Stop execution                    |
| `      | Syntax quoting | Like ' but qualify with namespace |
| ~      | Unquoting      | Start execution                   |

> If we quote something with ~\'~ or ~\`~ it will not be evaluated yet[11]. The difference between the two is that ~\`~ qualifies each expression with their full namespace. The ~\~~ symbol turns on evaluation inside of a ~\`~ again[11].

Reviewer: I’m not sure if it's very productive to think of macros as mainly starting and stopping execution of code pieces. It's just a function on code, i.e. S-Expressions. Yes, `'(some code)` is sort of "stopping" the execution of `(some code)` but I think it's more productive to think of quoting as literal quoting. As in »He said he's hungry« vs. »He said "He’s hungry"«.

Author: I see what you mean, but I'm not sure the value of quoting is conveyed if I just point out that it is being quoted. I think "stopped for later" would be more fitting. What do you think?

Reviewer: Well, as I said, starting and stopping execution is kind of part of the story, but it’s not what macros are _about_. Wrapping something in a thunk (`(fn [] something)`) is also "stopping execution" and calling a thunk is also "starting execution" but thunks have nothing to do with macros.

Author: ```suggestion
We can quote our code with ~'~ or ~`~ to tell Clojure(Script) not to evaluate it[11]. Quoting can be thought of as if we wrap our expression in literal quotes. Here is an example of quoting with JavaScript code: ~[1,2,3].sort()~ and ~"[1,2,3].sort()"~, where the last example isn't an expression anymore, but a String. In Clojure(Script) we can turn our quoted code back into an expression by using the ~~~ symbol[11].
The difference between the ~'~ and ~`~ symbol is that ~`~ qualifies each expression with their full namespace. So ~`map~ gets turned into ~cljs.core/map~ instead of just ~map~.
Syntax quoting is an important tool for writing macros as it allows us to to control how symbols are interpreted. We will see later how this is used to create macros that need to treat some symbols in a special way.
```

That thunk example explains it well. I now mention that it is about "control about how symbols are interpreted", though I don't 100% agree with that either.

*** What are the uses cases for macros?

> Macros provide so much freedom that they enable us to enhance not just our API but also the language itself. If the problem is beyond manipulating data, but rather about manipulating the functions, macros are a good fit.

Reviewer: "manipulating functions" -> "manipulating code"

Author: ```suggestion
Macros provide so much freedom that they enable us to enhance not just our API but also the language itself. If the problem is beyond manipulating data, but rather about manipulating code, macros are a good fit.
```

They allow a developer to provide their API exactly how they imagine it. That's why we will make use of them in our implementation.
However, while things like binding symbols to values under the hood can be used to make things easier for the developer, it can also cause confusion as developers have no idea where the symbol actually came from and can only assume that it works because a macro is used. That assumption gives them little information though. They can also lead to confusing error messages, as another step is added before the evaluation. And because of their freedom they can be complicated to implement. Lastly, as we're going to discuss later, macros aren't as easy to implement in ClojureScript as they are in Clojure. So macros must be used with care, even if they can be a tremendous help.
*** Which macros will be implemented?

> Because the structure of ~let~ acts as a dead-end, so too would our ~let~ like structure act as an end for our sequential composition. As such, a macro will be implemented that wraps the ~runner~ primitive and enhances it with the ~let~ like structure. Other macros might be added for easier implementation as well, but are not necessary.

Reviewer: I don't get what you mean by "dead-end" here. `let` can be further composed just as a custom monadic macro could be further composed. For example:

```clojure
(let [foo (let [bar 312] bar)]
  foo)

(monadic [x y
                  z (monadic [foo bar] foo)]
  z)
```

Author: ```suggestion
To get our ~let~ like structure a ~runner~ macro will be implemented that wraps the ~runner~ primitive and enhances it. Other macros might be added for easier implementation as well, but are not necessary.
```
Yea that didn't make sense. I wanted to say that once we use a ``runner``, we can't do ``(runner [a1 runner2])``. But that also isn't important here. I trimmed it down.

** How parallel and sequential composition interact
The developer should be able to use the API like they use the other tools of reacl-c. At the same time though, there needs to be a clear border between the parallel and sequential composition, as they're fundamentally different. To guarentee that, the API introduced the ~Prog~ and ~Bind~ types. While neither works with the other reacl-c tooling (to discourage incorrect usage), both contain an ~Item~.

> Taking a ~Prog~ (or ~Bind~) and turning it into an ~Item~ is simple, thanks to ~return~ and ~show~. Turning an ~Item~ into a ~Prog~ is also simple and can be done with ~return~,

Reviewer: Is this correct? `return` works both ways?

Author: ```suggestion
Taking a ~Prog~ (or ~Bind~) and turning it into an ~Item~ is simple, thanks to ~runner~ and ~show~. Turning an ~Item~ into a ~Prog~ is also simple and can be done with ~return~,
```
Nope meant to write ``runner``.

Functionality that could check if an ~Item~ will ever emit a ~commit~ (or other types), would be something to add in the future. Perhaps an additional keyword like ~:state:~ for the ~return~ function of the ~core~ namespace in reacl-c could be added to handle this case.

> It's worth mentioning that that at the borders further composition of the type that has been moved away from, isn't possible anymore. A ~runner~ returns an ~Item~ which from that point on can only be meaningfully parallely composed. Likewise, wrapping a ~Prog~ within a ~div~ with other ~Items~ is also meaningless. The developer needs to make a choice at those points if they really are done composing, in order to switch to the different type.

Reviewer: … can only be composed serially, right?

Author: Wouldn't parallel be correct because the ``runner`` returns an ``Item``, instead of a ``Prog``? That  returned ``Item`` also never emits a commit. Or did I mix something up?

Reviewer: ah yes, you’re right

Another use-cases is the conditional loading of data (from a server). Here the result does matter and we need to provide the possibility to react to it. Of-course, the developer could just add another continuation which reacts to it, but that is rather a hack. The ability to supply a continuation as the "body" would be a great fit, too.
*** Determining the best fit
Seeing how all three of our use-cases benefit from having the option to react to the result, the implementation that just displays the last ~Prog~ indefinitely or shows nothing - is of little use. Using the actions of ~reacl-c~ is nice, but causes unwanted complications. Giving developers the option to handle the result or ignore it, by passing a continuation as a body, allows for all use-cases to work and causes minimal overhead for the developer. This also mirrors the functionality of ~let~.

> This however raises the question what should happen if no continuation is supplied. Should the last Prog be shown indefinitely? From a user experience perspective it's expected that an action has a reaction, thus it makes sense to not show something indefinitely, but rather display nothing. Another benefit would be that sequential composition is cleaning it self up, after being done.

Reviewer: Another option is to make the final continuation mandatory.

Reviewer: such that "what should happen if no continuation is supplied" isn't a question that arises.

Author: ```suggestion
This however raises the question what should happen if no continuation is supplied.
Should the last Prog be shown indefinitely? From a user experience perspective it's expected that an action has a reaction, thus it makes sense to not show something indefinitely, but rather display nothing. Another benefit would be that sequential composition is cleaning itself up, after being done.
The issue of receiving no continuation could be avoided entirely, by always requiring a continuation. 
The downside to this would be a minor annoyance for developers, but makes sense for internal use, as less code is needed to implement the above behavior. 
The API will enable developers to not specify a continuation, but it’ll actually pass an empty function instead. 
```

This section could need a lot more work. I'd keep it at this for now.

** Why monads are a good fit
Using monads makes sense because they allow us to abstract the actual logic (e.g. waiting for a ~Commit~, combining ~Progs~) away from the developer while providing strong tools to combine our sequential steps.

> The required functions also are a great fit for our API. With the ~>>=~ operation we give the developer an easy way to further compose their sequential components. ~return~ requires us to make a new type, which we can in turn use to distinguish our sequential and parallel compositions from each other.

Reviewer: I’m not sure whether "distinguish our sequential and parallel compositions from each other" is what `return` does or should do. If it’s the case, we need more justification.

Author: ```suggestion
The required functions also are a great fit for our API. With the ~>>=~ operation we give the developer an easy way to further compose their sequential components. We also need to make a new type, which we can in turn use to distinguish our sequential and parallel compositions from each other. The ~return~ function then acts as a constructor for our new type. 
```
 👍

> Because we wait for a ~Commit~ to be emitted, we need to take asynchronicity into account. Thus we need the previously described tools to build up a blueprint and a specially labeled environment in which that blueprint is being executed in. This is much like monads in Haskell too, where monads can be executed in a ~do~ block.

Reviewer: what’s a blueprint in this context?

Author: ```suggestion
Because we wait for a ~Commit~ to be emitted, we need to take asynchronicity into account. That means our sequential composition needs a specially labeled environment in which that it can be executed in. This is much like monads in Haskell too, where monads can be executed in a ~do~ block.
```

** Tail Call Optimisation

> Neither Java[12] nor versions of JavaScript that we use in our browsers[13], feature Tail Call Optimization (TCO). Both langauges are used as host languages, Java for Clojure and JavaScript for ClojureScript. Due to the high amount of nested function calls it is however an important feature for a functional language. With a correct implementation of Tail Call Optimisation it is guaranteed that successive invocations of nested function calls (like our monadic bind) won't cause a stack overflow. It can also enable the use of recursion with our bind elements. Something that can enable infinitely repeating workflows. Burdening the developer to worry about depth of composition would be undesirable, as the goal is to create an easy to use API.

Reviewer: include a reference for TCO such as https://help.luddy.indiana.edu/techreports/TRNNN.cgi?trnum=TR19

Author: ```suggestion
Neither Java[12] nor versions of JavaScript that we use in our browsers[13], feature Tail Call Optimization (TCO)[x]. Both langauges are used as host languages, Java for Clojure and JavaScript for ClojureScript. Due to the high amount of nested function calls it is however an important feature for a functional language. With a correct implementation of Tail Call Optimisation it is guaranteed that successive invocations of nested function calls (like our monadic bind) won't cause a stack overflow. It can also enable the use of recursion with our bind elements. Something that can enable infinitely repeating workflows. Burdening the developer to worry about depth of composition would be undesirable, as the goal is to create an easy to use API.
```
I'd link to the paper in the bottom. I see this reference as a link to the origin of the idea. I only skimmed the paper but couldn't find the word "TCO" (important?). They do seem to talk about the concept on page 11. 

I could reference to [[13]](https://2ality.com/2015/06/tail-call-optimization.html) again, though.

Reviewer: That paper was just the first that came to my mind. I don’t think there’s a _definitive paper_ for TCO. Maybe ask Stefan if he knows a good reference.

> It is therefore important to add further abstraction to the bind logic, to allow us to implement some kind of TCO around it.

Reviewer: "further abstraction" is the wrong term. TCO or no-TCO doesn't manifest in the API itself. It's maybe "more implementation work".

Author: ```suggestion
It is therefore important to add code to our bind logic, to allow us to implement some kind of TCO around it.
```
👍

*** TCO in ClojureScript
While ClojureScript isn't offering TCO out-of-the-box for every function call, it does provide the ~loop~ and ~recur~ functions which do a locale rewrite of the code into a loop[14]. This allows for worry free function invocations, no matter the depth.

> Another way to gain TCO is to use what is called a trampoline. Instead of stepping deeper and deeper into nested function calls, the function is called once and the result, which is a function, is saved. Now for as long as the function returns another function, we will call the result. If a value is returned, we will stop and return that instead, breaking our ~loop~. Listing 18 shows a crude example. See [12] for a more detailed explanation.

Reviewer: "If a value is returned" -> "If a value that's not a function is returned" (functions are values)

Author: ```suggestion
Another way to gain TCO is to use what is called a trampoline. Instead of stepping deeper and deeper into nested function calls, the function is called once and the result, which is a function, is saved. Now for as long as the function returns another function, we will call the result. If a value is returned (that isn't a function), we will stop and return that instead, breaking our ~loop~. Listing 18 shows a crude example. See [12] for a more detailed explanation.
```

#+CAPTION: Example of trampolines in JS. ~optimized~ calls ~work~ initially and loops for as long as it returns a ~function~.
#+begin_src javascript
  function work(y) {
      return (y === 0) ? true : () => work(y-1);
  }

  function optimized(x) {
      let r = work(x);
      while (typeof r === 'function') r = r();
      return r;
  }
#+end_src
*** Custom TCO
While ~loop~ provides a perfectly fine way to get the benefits of TCO for synchronizing functions, in order to work with the asynchronous, action driven, approach that reacl-c uses, a custom implementation needs to be developed. We will revisit the concept of the trampoline again during our implementation.
* Implementation
Previously we have discussed which functions our API should offer. Now we will focus on implementing these functions and their associated features. Aside from delivering the necessary functionality, we will also talk about optimizing the ~runner~ function and implementing macros for ease of use.
** then

> Discussed earlier, the ~then~ function needs to compose steps together. The basic idea is trivial. Take a ~Prog~ and a ~continuation~ and return a ~Bind~, which is nothing but a container ~record~ type that holds both of these values. Listing 18 shows the definition of a ~record~ type as well as the ~make-bind~ function.

Reviewer: … making this a so called "free monad" (probably not important, just fyi)

#+CAPTION: Record definition of the ~bind~ type and a helper function to filter out incorrect values.
#+begin_src clojure
  (defrecord Bind [prog cont])

  (defn make-bind [prog cont]
    {:pre [(prog? prog)]}
    (->Bind prog cont))
#+end_src

> However, just getting a ~Prog~ every time would be of little use. A ~Prog~ just contains a single step. Things get interesting if we want to pass a ~Bind~, because we cannot just wrap the ~Bind~ again, as it already contains a ~Prog~.

Reviewer: Very nice section!

Reviewer: I still think it would be better to say (in the very beginning of the thesis) that `bind` is not a type of its own but a value constructor for the type `prog`.

[[./images/cps-transformation.png]]
This allows us to avoid having to flatten the ~Bind~ anywhere else, which makes showing the ~Item~ inside of the ~Bind~ trivial. It also guarantees that the order of execution will always be correct, thanks to deconstructing the passed ~Bind~ completely.

Now, inside of our ~then~ function we need to handle both cases. For this we differentiate between a ~then~ call where A: a ~Prog~ is passed or B: where a ~Bind~ is passed. If a ~Prog~ is passed, we just wrap the parameters and return a ~Bind~. If however a ~Bind~ is passed, we do our CPS-transformation. See listing 19 for the previously described code.
#+CAPTION: Definition of the ~then~ function. The CPS-transformation can be seen in line 7, as the previous ~Bind~ and new ~Bind~ are first deconstructed and then reconstructed into a new ~Bind~.
#+begin_src clojure -n 1
  (defn then [prog cont]
    {:pre [(or (bind? prog) (prog? prog) (c/item? prog))
	   (fn? cont)]
     :post [(bind? %)]}
    (if (bind? prog)
      (make-bind (bind-item prog) (fn [x] (then ((bind-continuation prog) x) cont)))
      (make-bind (if (c/item? prog) (make-prog prog) prog) cont)))
#+end_src

> To give the developer feedback in case they make an error, we add ~:pre~ and ~:post~ annotations, which let ClojureScript know to check the types that come into and out of our function. In this case we say that the ~prog~ can be a ~Prog~, ~Bind~ or even an ~Item~. The ~cont~ parameter needs to be a function and the result of our operation should always return a ~Bind~.

Reviewer: coercing an item to a prog may be nice in a real-world library but it adds no value for your thesis imo.

Author: ```suggestion
To give the developer feedback in case they make an error, we add ~:pre~ and ~:post~ annotations, which let ClojureScript know to check the types that come into and out of our function. In this case we say that the ~prog~ can be a ~Prog~ (or it's subtype ~Bind~). The ~cont~ parameter needs to be a function and the result of our operation should always return a ~Bind~.
```

I'd leave the behavior in the code, but remove the text mentioning it. What do you think?

	  (show inner))
	 (fn [[outter st] ac]
	   (if (and (commit? ac) (bind? st))
	     (core/return :state [outter ((bind-continuation st) (commit-payload ac))])
	     (core/return :action ac))))))))
#+end_src
*** Basic Functionality
The goal of the ~runner~ is to hold and display what is inside of a ~Bind~ (or ~Prog~). While it does this, it wraps the ~Bind~ and waits for a ~Commit~ which will trigger it to call the ~continuation~ of the ~Bind~.
To understand this better we're going to focus on lines 7-14 of listing 21. First notice the ~core/handle-action~ call. This function takes two parameters. First an ~Item~ to display, second a ~function~ to call once the ~Item~ emits something. The function that will be called can be seen at the bottom on lines 11-14. We don't directly pass an ~Item~ however. We pass a ~focus~ function. This is done to ~core/focus~ the state on a specific part. Namely limiting the implicitly passed state to what is outside of the runner, instead of leaking the bookkeeping state of the runner downwards. This function again takes two parameters. First a lens (function of two arities) and second an ~Item~. We will talk more about the lense in the next chapter.

> The second parameter is the in our API defined ~show~ function, which just takes either ~Bind~, ~Prog~ or ~Item~ and unwraps it to an ~Item~ again.

Reviewer: :D This sentence doesn't work in English. "The second parameter is the `show` function defined in our API."

Author: ```suggestion
The second parameter is the ~show~ function defined in our API, which just takes either ~Bind~, ~Prog~ or ~Item~ and unwraps it to an ~Item~ again.
```

Ups, hehe.

*** Lens
The previously mentioned lens in line 9 is needed to stop leaking the bookkeeping for our tail call optimization, downards into our ~Item~.
Lenses are a popular mechanism in functional programming to, on one side, restrict the available information, while allowing changes from the restricted side to change the whole, as well.
It does this by providing two functions: Yanker and shover. The yanker grants access to parts of the whole. If parts of that whole are changed, the shover is called to marry that part again with the whole, so the update can trickle upwards again.
Our lens is a "first lense", because it restricts access to anything but the ~first~ element. This is relevant because when state is passed around in reacl-c, it usually comes in the form of a list where the ~first~ element is the outter and the second element is the inner state.
The code for the ~first-lens~ can be found in listing 22.

> #+CAPTION: Elements of the list are destructured into ~first~ and ~rest~, where on the yanker side ~rest~ is discarded and on the shover side ~first~ is ignored in favor of ~small~. ~small~ represents the previously passed ~first~, which now has been updated by the restricted side.

Reviewer: Maybe discuss lenses further up as a prerequisite(?)

#+begin_src clojure
  (defn first-lens
    ([[first & _]]
     first)
    ([[_ & rest] small]
     (vec (cons small rest))))
#+end_src

> Because we don't want the internal state of our ~runner~ to leak, but we do want the state surrounding the ~runner~ to be passed down, we ignore the inner state (bookkeeping) and pass down the outter state (implicitly passed state).

Reviewer: Maybe mention again that `local-state` introduces local state, therefore necessitating the use of `(focus first ...)`

Author: ```suggestion
Because we don't want the internal state introduced by our ~local-state~ call (inside of our ~runner~) to leak, but we do want the state surrounding the ~runner~ to be passed down, we ignore the inner state (bookkeeping) and pass down the outer state (implicitly passed state).
```
Does this make it clearer?

> Our lense is a function with multiple arities. That means it can take different amounts of parameters. Ours being of arity of two, means it has two different signatures. A signature where it gets one parameter and a signature where it takes two parameters.

Reviewer: The function is not of arity two but of arities 1 and 2.

Author: ```suggestion
Our lens is a function with multiple arities. That means it can take different amounts of parameters. Ours being of arity 1 and 2, means it has two different signatures. A signature where it gets one parameter and a signature where it takes two parameters.
```
👍

To restrict access, so when it is called from the perspective of the child, the signature with one parameter is called. That is our yanker. Here the passed parameter is destructured and everything besides the ~first~ element is ignored. That ~first~ element is then returned. On changes to the state from within our child, the signature with two parameters is called. That is our shover. The change being the second parameter, here called ~small~. We again destructure the argument but now ignore the previously named ~first~ and instead access the previously ignored ~rest~. All that is left to do is to combine them with ~cons~ and return them as a list.
*** Implementing Tail Call Optimization
Previously we discussed what TCO is and why we need it. Now it is time to take a deeper look at how we implement it by examining lines 3-6 and lines 11-14 in listing 21.

> Like mentioned earlier ClojureScript already provides TCO with the ~loop~ macro, unfortunately we can't take advantage of it because of the asynchronous nature of reacl-c. This is not a problem as we can use the previously explained concept of trampolines to make our own TCO.

Reviewer: "asynchronous nature of reacl-c" is not quite correct. reacl-c works very synchronously. I know what you mean, but I'd phrase it differently for readers who are unfamiliar with reacl-c. Maybe just: `loop` and `recur` are for recursion on functions. That's now quite what we're here for, though. We need recursion on Progs. We therefore cannot simply use `loop` and `recur` as is …

Author: ```suggestion
As mentioned earlier ClojureScript already provides TCO for functions with the ~loop~ and ~recur~ macros. We however need recursion on our ~Progs~, because of that we cannot simply use ~loop~ and ~recur~. This is not a problem though, as we can use the previously explained concept of trampolines to make our own TCO.
```
👍

#+CAPTION: Code for the ~show~ function.
#+begin_src clojure
  (defn show
    [x]
    {:post [(c/item? %)]}
    (cond
      (prog? x) (prog-item x)
      (bind? x) (prog-item (bind-item x))
      (c/item? x) x
      :else (c/fragment)))
#+end_src
** Macros
The available primitives already enable the developer to craft elaborate sequential compositions, however they aren't able to provide the ~let~ like usage on their own. For this reason, we will implement macros for our API. Only a macro version of the ~runner~ is required, but it makes more sense to enhance the ~then~ primitive first and later wrap the ~runner~ with something that takes advantage of the ~then~ macro. Before we dive into the macros themselves, we first need to take a look at how macros work in ClojureScript.
*** TODO Setup

> Both Clojure and Clojurescript have access to macros, though implementing one is more complex in the later. This is because of the compilation process. Macros are run during the compile process of ClojureScript. In short, this means the compilation process needs to be kept in mind when writing ClojureScript macros. You cannot, for instance, put macros in a ~.cljs~ file.

Reviewer: In short: Macros are always expanded by Clojure, even though they may produce ClojureScript code.

So a macro on Clojure:
```
ClojureCode -> ClojureCode
```

is run by Clojure. And a macro on ClojureScript:

```
ClojureScript -> ClojureScript
```

is also run by Clojure.

Author: ```suggestion
Both Clojure and ClojureScript have access to macros, though implementing one is more complex in the later. This is because macros are always expanded by Clojure, even if they may produce ClojureScript code. This means Clojure is always involved, even in pure ClojureScript projects. This means the compilation process needs to be kept in mind when writing ClojureScript macros. You cannot, for instance, put macros in a ~.cljs~ file.
```
I like your "in short" explanation and jammed it in there. I still mention "the compilation process", but I can change that if it's incorrect etc.

*** runner
With our ~then~ we have greatly improved the desired developer experience, however, the result still returns a ~Bind~. That is fine, as we might want to further compose this. However, we still need to wrap our ~then~ expression with a ~runner~ to run it. To further simplify the experience we will create another macro, this time for the ~runner~ function to give developers the option to do everything within a single call.
For this we will simply wrap our ~then~ macro with our primitive ~runner~ from the ~bind~ namespace. Again we syntax quote our call and undo the quote for our values. To add even more convenience our ~runner~ macro has an arity of two. If the developer is not interested in supplying an ~end-expr~ function, we will pass an empty function into the ~then~ for them. The resulting code is simple but works exactly as we want it to and can be seen in listing 29.
#+CAPTION: Definition of the ~runner~ macro.
#+begin_src clojure
  (defmacro runner
    ([x]
     `(runner ~x (fn [])))
    ([x y]
     `(code.bind/runner (then ~x  ~y))))
#+end_src
** Limitations

> While the current implementation achieves what it set out to do, some compromises had to be made. First, as mentioned earlier, it is being relied on the fact that the developer actually emits a ~Commit~ in what they label a ~Prog~. There is currently no logic to make sure that the developer is forced too or reminded if they aren't. Another limitation is that the current API offers only primitives and our two macros, but none of the deep functionality which is found in reacl-c for the parallel composability. So functions that map, filter etc. over sequential compositions are not included. There is also no error handling for sequential composition. Developers need to handle errors by hand in the continuation of the next ~Bind~, as there is no ~Error~ sub-type of ~Commit~. Lastly, there is no way to terminate early, like with a ~Maybe~ monad.

Reviewer: > it is being relied on the fact that the developer actually emits a ~Commit~ in what they label a ~Prog~

A nice way to solve this is to offer only a primitive `wait-for-action : Item -> Predicate -> Prog`. `wait-for-action` takes an item and listens for actions. When an action satisfies `Predicate` it returns a `Commit`. The `Commit` record itself could then be an internal implementation detail of the library.

Reviewer: > So functions that map, filter etc. over sequential compositions are not included.

These could simply be written via `bind` and `return`. 

In Haskell: `fmap f x = x >>= (return . f)`

Author: I like the idea of ``wait-for-action`` and I agree that ``fmap`` would be trivial. I see both as a v1.1 kinda deal, therefor mention it under "Limitations".

But, not just a separate type for terminating early would be of interest. Having the ability to communicate that an error has occurred and then being able to handle that (maybe even in different steps of severity) would be also beneficial. A possible scenario would be a lot of time passing since step 1 and step 2 and the user gets logged out. Instead of having to handle the error either outside of the ~runner~ or inside of every single step, we could dedicate a space inside of the ~runner~ for exactly that.

> Besides new types, another improvement would be to force a developer to ~Commit~ something inside of a ~Prog~. The implementation for this is unclear, however the feature is of interest as that could be a major source of bugs.

Reviewer: See comment above about `wait-for-action`
